### DISCUSSION TITLE
Spans with the same text but different labels share the same extension value

### DISCUSSION THREAD

**Discussion by polkaYK** [explosion/spaCy#8901] [Help: Other Questions] [answered]
Hello. I don't know if it's a bug or just something that I do not fully understand. We've found this issue at work, but here I prepared a little toy example that can reproduce its behavior.

I have a doc with a bunch of spans. Some spans can share the same text, but have different labels (this comes from auto-generated BRAT annotation that goes along with the text).
```
import spacy
nlp = spacy.load('en_core_web_sm')
doc = nlp('Mary had a little lamb')

span1 = Span(doc, start=0, end=2, label="LABEL_1")
span2 = Span(doc, start=0, end=2, label="LABEL_2")
span3 = Span(doc, start=3, end=4, label="LABEL_3")
```
At first I thought that span1 and span2 are different objects, since they have different ids, hashes and are not equal to each other
```
print(id(span1)) # 139725572848320
print(id(span2)) # 139725495655664

print(hash(span1)) # 3674900633150140320
print(hash(span2)) # -3886871937572295988

print(span1 == span2) # False
```
The problem arises when I want to create a span extension that will store a list of objects connected to each span
```
from spacy.tokens.span import Span
Span.set_extension('relations', default=None) # note, that default value is None, not empty list

span1._.relations = []
span1._.relations.append(f"relation_with_label: {span1.label_}")
print(span1._.relations) # ['relation_with_label: LABEL_1']
```
The thing that surprised me was that span2 that shared the same text was also linked to the same list. Span3's "relations" is still None.
```
print(span2._.relations) # ['relation_with_label: LABEL_1']
print(span3._.relations # nothing happens

print(id(span1._.relations)) # 139726319670464
print(id(span2._.relations)) # 139726319670464
``` 
So, my question is, why spans that share the same text but are considered different python objects also share the same extension list? Is there a way to implement the extension when both span1 and span2 hold their own list of "relations"? 

## Info about spaCy

- **spaCy version:** 3.1.1
- **Platform:** Linux-5.8.0-63-generic-x86_64-with-glibc2.31
- **Python version:** 3.9.4
- **Pipelines:** en_core_web_sm (3.1.0)

**Comment by polm** [comment, marked as answer]
This was a bit of a surprise to me but I confirmed it is happening and figured out why. 

When you create a Span extension, the data is actually saved on the Doc, in `Doc.user_data` (a dict). The key is a tuple that includes the field name and the span start and end, but not the span object id or other info. So two spans with the same start and end will have the same data. 

I guess this is a design decision, but it is surprising that items without object identity share data, and we should probably highlight this more in the docs. On the other hand, I'm having a hard time imagining concrete cases where it would make sense for spans over the same text to have different user data. Could you tell us some more about your use case?

Also, though it's a bit awkward, I think you can work around this in various ways like setting a dictionary in the attribute where the key is the object id, or using a custom getter to wrap similar behavior.

**Reply by polkaYK** [reply]
We work with texts of contracts, that once were docx documents with complex structure (sometimes all text of a document is in a big table, or different parts of a contract are in different cells of such table). Since this table structure is important for further work, some third-party api generates BRAT annotation of this document, which aims to store all this table relations. Sometimes it's possible, that the same spans (accurate up to a whitespaces) appear with different tags in this annotation. For example (T185 and T184 are different entities in annotation, but store the same text):
```
T185	CELL 16860 16889	 
T184	ROW 16860 16912	
```
We then create Span objects from each of these records with BRAT tag as a label and want to store all their relations with other elements in a list. As you can see, doing this through extension was not the best choice. 
I want to emphasize, that these spans with same texts, but different labels are not our special design choice (as I am aware of), it's something, that occurs sometimes because of complex document structure and automatic annotation.

**Reply by polm** [reply]
I'm not entirely sure I follow your annotations (what are the third and fourth columns? character offsets?), but it sounds like maybe you could fix this by storing labels as a list rather than as a single string value, so that you could have multiple annotations at the same location? 

It sounds like you don't need to store conflicting data, just that you might have more than one value for a given span.
